tap/passive_tapper.go

```diff
@@ -198,8 +201,41 @@ func startMemoryProfiler() {
 	}()
 }
 
+func closeTimedoutTcpStreamChannels() {
+	cpuCount := runtime.NumCPU()
+	// +1 Stats tracker Goroutine
+	// +1 Main packet capture loop
+	baseGoroutineCount := runtime.NumGoroutine() + 2
+	for {
+		runtime.Gosched()
+		streams.Range(func(key interface{}, value interface{}) bool {
+			streamWrapper := value.(*tcpStreamWrapper)
+			stream := streamWrapper.stream
+			// `diff` tries to stabilize towards 0.
+			// `diff` means the number of currently running extra Goroutines mainly spawned for dissectors.
+			n := runtime.NumGoroutine()
+			diff := n - baseGoroutineCount
+			if diff < 1 {
+				diff = 1
+			}
+			// `metric` tries to stabilize arround 1.0.
+			// `metric` means what percentage of the load can be actually handled by the given resources.
+			metric := float64(cpuCount) / float64(diff)
+			dynamicStreamChannelTimeoutMs := int(float64(baseStreamChannelTimeoutMs) * metric)
+			dynamicStreamChannelTimeoutNs := dynamicStreamChannelTimeoutMs * 1000000
+			if !stream.isClosed && stream.superIdentifier.Protocol == nil && time.Now().After(streamWrapper.createdAt.Add(time.Duration(dynamicStreamChannelTimeoutNs))) {
```

Comment:
What does it mean that `stream.superIdentifier.Protocol == nil`?
If we have `stream.superIdentifier.Protocol` non-nil for an inactive stream, will it ever close?

Reply:
I think the algorithm can be simplified by setting the maximum allowed nil-protocol-time to a const (the argument to `After`).
Do your tests lead to a different conclusion?

Reply:
@nimrod-up9 if `stream.superIdentifier.Protocol == nil` is `true` it means that the TCP stream is not yet identified by any protocols. If it's `false` it means the TCP stream is identified by a protocol.

If `stream.superIdentifier.Protocol` is non-nil then yes the TCP stream is not dropped. The other protocols break their read loop as it's explained in here https://github.com/up9inc/mizu/pull/260#discussion_r703020684

`stream.superIdentifier.Protocol == nil` criteria is there because the AMQP streams are long lasting and shouldn't be dropped. Please see https://github.com/up9inc/mizu/pull/260#issuecomment-913079684

Reply:
@nimrod-up9 the algorithm cannot be simplified because it scales vertically and the timeout is rearranged based on the load. 

The formula tries to answer this question: "Do we have enough resources to handle that amount Goroutines?" If the answer to this question is no then it drops the **old, unidentified TCP streams**.

There is a room for the improvement in the formula though. Please see;
- https://github.com/up9inc/mizu/pull/260#issuecomment-913082888
- https://github.com/up9inc/mizu/pull/260#issuecomment-913126666